source/extensions/priv/server/ntds_decrypt.c

```diff
@@ -0,0 +1,194 @@
+* @param output Pointer to the string we are outputting the result to
+*/
+void bytes_to_string(LPBYTE data, int length, LPSTR output){
+	for (int i = 0; i < length; i++){
+		sprintf(output + (i << 1), "%02X", ((LPBYTE)data)[i]);
+	}
+}
+
+/*!
+* @brief Decrypt an  encrypted LM or NT Hash.
+* @param encryptedNTLM Pointer to an encryptedhash struct for the LM or NT hash we wish to decrypt.
+* @param pekDecrypted Pointer to a decryptedPEK structure that holds our decrypted PEK
+* @param hashString Pointer to the string where we will store the decrypted hash
+* @param rid DWORD representing the Relative ID(RID) of the account
+* @returns Indication of sucess or failure.
+*/
+BOOL decrypt_hash(struct encryptedHash *encryptedNTLM, struct decryptedPEK *pekDecrypted, char *hashString, DWORD rid){
+	BOOL cryptOK = FALSE;
+	BYTE encHashData[NULL_TERIMNATED_HASH_LENGTH] = { 0 };
+	BYTE decHash[NULL_TERIMNATED_HASH_LENGTH] = { 0 };
+
+	memcpy(&encHashData, &encryptedNTLM->encryptedHash, HASH_LENGTH_BYTES);
+	cryptOK = decrypt_rc4(pekDecrypted->pekKey, encryptedNTLM->keyMaterial, encHashData, 1, HASH_LENGTH_BYTES);
+	if (!cryptOK){
+		return FALSE;
+	}
+	cryptOK = decrypt_hash_from_rid(encHashData, &rid, decHash);
+	if (!cryptOK){
+		return FALSE;
+	}
+	bytes_to_string(decHash, HASH_LENGTH_BYTES, hashString);
+	return TRUE;
+}
+
+/*!
+* @brief Wraps SystemFunction025 which decrypts a hash using the RID
+* @param encodedHash Pointer to a byte array containing the encrypted hash
+* @param rid Pointer to a DWORD containing the Relative ID(RID)
+* @param decodedHash Pointer to where we store the decrypted hash
+* @returns Indication of sucess or failure.
+*/
+BOOL decrypt_hash_from_rid(LPBYTE encodedHash, LPDWORD rid, LPBYTE decodedHash){
+	typedef NTSTATUS(__stdcall *PSYS25)(IN LPCBYTE data, IN LPDWORD key, OUT LPBYTE output);
+	HMODULE hAdvapi = LoadLibrary("advapi32.dll");
+	if (hAdvapi == NULL){
+		return FALSE;
+	}
+	PSYS25 decryptFromRID = (PSYS25)GetProcAddress(hAdvapi, "SystemFunction025");
+	if (decryptFromRID(encodedHash, rid, decodedHash) != 0){
+		return FALSE;
+	}
+	return TRUE;
+}
+
+/*!
+* @brief Splits up an encrypted LM or NT hash history and decrypts each stored hash
+* @param encHashHistory Pointer to the byte array containing the hash history record
+* @param sizeHistory The size of the history record
+* @param pekDecrypted Pointer to a decryptedPEK structure that holds our decrypted PEK
+* @param rid DWORD representing the Relative ID(RID) of the account
+* @param accountHistory Pointer to a string wherewe store all the decrypted historical hashes
+* @param historyCount Pointer to n integer where we store a count of the historical hashes
+* @returns Indication of sucess or failure.
+*/
+BOOL decrypt_hash_history(LPBYTE encHashHistory, size_t sizeHistory, struct decryptedPEK *pekDecrypted, DWORD rid, char *accountHistory, int *historyCount){
+	BOOL cryptOK = FALSE;
+	int sizeHistoryData = (int)sizeHistory - 24;

```

Comment:
Even if you use unsigned be careful with this -24, the sizeHistoryData var is used later for memcpy size